catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,186 @@
+
+    private int pageSize;
+
+    //Set to true when a query returns zero results.
+    private boolean isExhausted;
+
+    private BlockingQueue<Result> queriedResultsBuffer;
+
+    /**
+     * QueryResultPaginator takes in the CatalogFramework object and a Query Request and
+     * queries for the number of results needed. This is done using the pageSize specified
+     * in the Query object. The Paginator will return a collection of results of the
+     * specified page size.
+     * <p>
+     * The startIndex is the first index in the results gathered from the CatalogFramework
+     * <p>
+     * The endIndex is the last index in the results gathered from the CatalogFramework, which
+     * may not be reached if the pageSize is smaller than the difference between the endIndex
+     * and startIndex.
+     * <p>
+     * The resultsRemaining are the total results not yet returned in a collection of the size
+     * of a page. This value is tracked through the hasNextPage() method to inform the client that
+     * there are valid results to continue querying.
+     *
+     * @param catalogFramework
+     * @param query
+     */
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryImpl query) {
+
+        Validate.notNull(catalogFramework);
+        Validate.notNull(query);
+
+        this.catalogFramework = catalogFramework;
+        this.queryCopy = copy(query);
+
+        isExhausted = false;
+        pageSize = queryCopy.getPageSize();
+        indexOffset = queryCopy.getStartIndex();
+
+        Validate.inclusiveBetween(1,
+                Integer.MAX_VALUE,
+                indexOffset,
+                "Query request start index must be greater than zero");
+
+        queriedResultsBuffer = new LinkedBlockingDeque<>();
+    }
+
+    // TODO: 6/27/17 Move this method over to QueryResultIterable inner class
+    /**
+     * If there are pages beyond the current page, returns true.
+     *
+     * @return True if there are results beyond the current page
+     */
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (notEmpty()) {
+             return true;
+         }
+
+         if (isExhausted) {
+             return false;
+         }
+
+         //Buffer is empty, but query might not be exhausted.
+         fetchNext();
+         return notEmpty();
+    }
+
+    protected boolean notEmpty() {
+        return !queriedResultsBuffer.isEmpty();
+    }
+
+    /**
+     * Catalog Framework queried for results until current page is full of results
+     * or endIndex is reached.
+     *
+     * @return Collection of results
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        while (!isExhausted && getBufferSize() < pageSize) {
+            fetchNext();
```

Comment:
I agree it can be cleaned up quite a bit, as it's pretty convoluted and does not have a great deal of focus.